### DISCUSSION TITLE
Suggestions of coupling interface functionality to Auxkernels or BCs or UserObject?

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#24654] [Q&A General] [answered]
Hi all, I'm trying to find the appropriate approach to do some calculations at the interfaces using interface quantities and enforce the outcome (velocity or displacement) as boundary condition along the interface nodes strongly before solving the system, I wonder whether I should start with AuxKernels or BCs or UserObject? Appreciate your suggestions

**Comment by GiudGiud** [comment]
Hello

BCs, DirichletBCs, can set values strongly (using the `preset` parameter as true), which even reduces the size of the nonlinear system.

I would start there. 
AuxKernels are not made for boundary work, though there are some options there.
InterfaceKernels are likely where you want to work if these boundaries are true interfaces (using variables defined on both sides)

Guillaume

**Reply by chunhuizhao478** [reply]
Thanks Guillaume, I think InterfaceKernels could weakly set the displacement across the interface. That could be something I want to try. But if I want to use BC system along interface (that is, to write all the calculations inside BC object with the interface information available)  is there a workaround available? Thanks.

**Reply by GiudGiud** [reply]
You can have a look at MatchedValueBC see how it gets information from the other variable. 
A BC has to be external to a variable's domain. The domain for that variable must end at the BC.
At least that's true for finite volume. In finite element, I've never dealt with that.

**Reply by lindsayad** [reply]
You could use a nodal BC to impose some strong constraint on a nodeset interior to the a finite element variable's domain in theory, but I've never seen someone do it in practice

**Reply by chunhuizhao478** [reply]
Thanks. I choose to use InterfaceKernel, it almost be able to achieve what I want, however, If I want to put all the calculations inside the object inherited from InterfaceKernel, I have to **update some quantities and save them for the next time step use**. I'm struggling to find a way to do it, neither material property or aux variables seems not working.. I know Kernels have the "save_in" option to save residual and output as aux variable, should I modify that part or there is a better approach

**Reply by GiudGiud** [reply]
You should use stateful material properties
OR 
auxiliary variables and the coupledValueOld routine to retrieve the previous step value of the variable

**Reply by chunhuizhao478** [reply]
Thanks, but the question is, I cannot update them inside the kernel :(

**Reply by chunhuizhao478** [reply, marked as answer]
Okay, I think couple InterfaceMaterial (do all calculations, updates using material property) and InterfaceKernel (pass material property) could do the trick

**Reply by GiudGiud** [reply]
yes, to update those quantities you need to do it in a Material object for a property or an Auxkernel for an auxiliary variable

**Reply by chunhuizhao478** [reply]
Yes, in my case, AuxKernel seems not suitable for retrieving quantities across interface, so Material object is better
